modules/network.go

```diff
@@ -81,218 +84,349 @@ func (n *network) startHealthcheck() {
+		}
 	}
-	// Also update latest block number with reference block if it's higher
-	if referenceBlock > n.latestBlockNumber {
-		n.latestBlockNumber = referenceBlock
+	return true
+}
+
+// getLatestHealthyBlock returns the highest block number among healthy providers
+// Falls back to warning providers if no healthy providers are available
+func (n *network) getLatestHealthyBlock() int64 {
+	var latestBlockFromHealthy int64
+	var latestBlockFromWarning int64
+	var latestBlockFromUnhealthy int64
+
+	// Single pass through providers to track latest blocks by status
+	for _, provider := range n.Providers {
+		history := provider.BlockHistory()
+		if len(history) == 0 {
+			continue
+		}
+
+		lastBlock := history[len(history)-1].blockNumber
+		switch provider.healthStatus {
+		case Healthy:
+			if lastBlock > latestBlockFromHealthy {
+				latestBlockFromHealthy = lastBlock
+			}
+		case Warning:
+			if lastBlock > latestBlockFromWarning {
+				latestBlockFromWarning = lastBlock
+			}
+		case Unhealthy:
+			if lastBlock > latestBlockFromUnhealthy {
+				latestBlockFromUnhealthy = lastBlock
+			}
+		}
 	}
 
-	if providerBlockNumber+n.BlockLagLimit < referenceBlock {
-		n.logger.Warn("Provider is lagging behind",
-			zap.String("provider", providerName),
-			zap.String("network", n.Name),
-			zap.Int64("provider_block_number", providerBlockNumber),
-			zap.Int64("reference_block_number", referenceBlock),
-			zap.String("machine_id", n.machineID))
-		provider.markWarning()
-	} else {
-		provider.markHealthy(n.HCThreshold)
+	// Return highest block number, prioritizing by health status
+	if latestBlockFromHealthy > 0 {
+		return latestBlockFromHealthy
 	}
+	return latestBlockFromWarning
 }
 
-func (n *network) sendLatestBlockMetric(providerName string, statusCode int, healthStatus string, providerBlockNumber int64) {
-	n.PrometheusClient.HandleLatestBlockMetric(&prom.PromLatestBlockMetricData{
-		Network:        n.Name,
-		Provider:       providerName,
-		ResponseStatus: statusCode,
-		HealthStatus:   healthStatus,
-		BlockNumber:    providerBlockNumber,
+func (n *network) sendHealthCheckMetric(providerName string, healthStatus string) {
+	n.PrometheusClient.HandleHealthCheckMetric(&prom.PromHealthCheckMetricData{
+		Network:      n.Name,
+		Provider:     providerName,
+		HealthStatus: healthStatus,
 	})
 }
 
-func (n *network) getCheckedProviderHCList(providerName string) ([]healthCheckEntry, bool) {
-	n.healthCheckListMutex.RLock()
-	defer n.healthCheckListMutex.RUnlock()
-	values, ok := n.CheckedProviders[providerName]
-	return values, ok
-}
+func (n *network) getLatestBlockNumber(httpUrl string, headers map[string]string, ac auth.IAuthClient) (int64, HealthStatus, error) {
+	var lastErr error
+	var lastHealthStatus HealthStatus = Unhealthy
+
+	// Layer 1: Handle attempts
+	for attempt := 0; attempt < n.RequestAttemptCount; attempt++ {
+		resBytes, statusCode, err := n.tryBlockNumberRequestWithTimeout(httpUrl, headers, ac)
+		if err != nil {
```

Comment:
```suggestion
		if err != nil && lastHealthStatus == Unhealthy{
```

IMO if we got an error in processBlockNumberResponse which leaded to an updated status, better to hold that message, instead of overwritting it over next attempt, so returned status and error message belong to the same iteration of the loop

Reply:
i removed this function completely.